File: yinwei/contracts/v1/js/coordinate_frame_v1.mjs
//! CoordinateFrameV1 — right-handed Y-up, −Z forward, metres.

export const QUAT_DEGENERATE_NORM = 1e-12;
export const ZERO_DISTANCE_M = 1e-9;
export const HORIZONTAL_POLE_M = 1e-8;

export function vec3(x, y, z) {
  return { x, y, z };
}

export function quat(w, x, y, z) {
  return { w, x, y, z };
}

export const IDENTITY_QUAT = quat(1, 0, 0, 0);
export const HRTF_FORWARD_FALLBACK = vec3(0, 0, -1);

export function vecLength(v) {
  return Math.sqrt(v.x * v.x + v.y * v.y + v.z * v.z);
}

function cross(a, b) {
  return vec3(
    a.y * b.z - a.z * b.y,
    a.z * b.x - a.x * b.z,
    a.x * b.y - a.y * b.x,
  );
}

function add(a, b) {
  return vec3(a.x + b.x, a.y + b.y, a.z + b.z);
}

function sub(a, b) {
  return vec3(a.x - b.x, a.y - b.y, a.z - b.z);
}

function scaled(v, s) {
  return vec3(v.x * s, v.y * s, v.z * s);
}

export function quatNorm(q) {
  return Math.sqrt(q.w * q.w + q.x * q.x + q.y * q.y + q.z * q.z);
}

export function normalizeQuat(q) {
  const n = quatNorm(q);
  if (n < QUAT_DEGENERATE_NORM) return IDENTITY_QUAT;
  return quat(q.w / n, q.x / n, q.y / n, q.z / n);
}

export function inverseQuat(q) {
  const n = normalizeQuat(q);
  return quat(n.w, -n.x, -n.y, -n.z);
}

export function rotateVec3(q, v) {
  const n = normalizeQuat(q);
  const qvec = vec3(n.x, n.y, n.z);
  const t = scaled(cross(qvec, v), 2);
  return add(add(v, scaled(t, n.w)), cross(qvec, t));
}

export function wrapAzimuthDeg(deg) {
  let az = deg;
  while (az > 180) az -= 360;
  while (az <= -180) az += 360;
  return az;
}

export function shortestAzimuthDeltaDeg(fromDeg, toDeg) {
  let d = toDeg - fromDeg;
  while (d > 180) d -= 360;
  while (d < -180) d += 360;
  return d;
}

export function lerpAzimuthDeg(fromDeg, toDeg, t) {
  return wrapAzimuthDeg(fromDeg + shortestAzimuthDeltaDeg(fromDeg, toDeg) * t);
}

export function sphericalToLocal(p) {
  const az = (p.azimuthDeg * Math.PI) / 180;
  const el = (p.elevationDeg * Math.PI) / 180;
  const d = p.distanceM;
  const ce = Math.cos(el);
  return vec3(d * Math.sin(az) * ce, d * Math.sin(el), -d * Math.cos(az) * ce);
}

export function localToSpherical(p) {
  const r = vecLength(p);
  if (r < ZERO_DISTANCE_M) {
    return { azimuthDeg: 0, elevationDeg: 0, distanceM: 0 };
  }
  const h = Math.sqrt(p.x * p.x + p.z * p.z);
  if (h < HORIZONTAL_POLE_M) {
    return {
      azimuthDeg: 0,
      elevationDeg: p.y >= 0 ? 90 : -90,
      distanceM: r,
    };
  }
  const el = (Math.asin(Math.min(1, Math.max(-1, p.y / r))) * 180) / Math.PI;
  const az = wrapAzimuthDeg((Math.atan2(p.x, -p.z) * 180) / Math.PI);
  return { azimuthDeg: az, elevationDeg: el, distanceM: r };
}

export function hrtfUnitFromLocal(local) {
  const r = vecLength(local);
  if (r < ZERO_DISTANCE_M) return HRTF_FORWARD_FALLBACK;
  return scaled(local, 1 / r);
}

export function worldToListenerLocal(world, listener) {
  const rel = sub(world, listener.worldPosition);
  return rotateVec3(inverseQuat(listener.orientation), rel);
}

export function listenerLocalToWorld(local, listener) {
  return add(listener.worldPosition, rotateVec3(listener.orientation, local));
}

export function geometricDistanceM(local) {
  const r = vecLength(local);
  return r < ZERO_DISTANCE_M ? 0 : r;
}

File: yinwei/apps/yinwei_player/lib/contracts/coordinate_frame_v1.dart
/// CoordinateFrameV1 — Spatial Scene Domain axes and conversions.
///
/// Isolated from playback / HRTF / UI. Not wired into runtime in Phase 1A.
library;

import 'dart:math' as math;

const double kQuatDegenerateNorm = 1e-12;
const double kZeroDistanceM = 1e-9;
const double kHorizontalPoleM = 1e-8;

class Vec3V1 {
  const Vec3V1(this.x, this.y, this.z);

  final double x;
  final double y;
  final double z;

  double get length => math.sqrt(x * x + y * y + z * z);

  Vec3V1 operator +(Vec3V1 o) => Vec3V1(x + o.x, y + o.y, z + o.z);

  Vec3V1 operator -(Vec3V1 o) => Vec3V1(x - o.x, y - o.y, z - o.z);

  Vec3V1 scaled(double s) => Vec3V1(x * s, y * s, z * s);

  static const zero = Vec3V1(0, 0, 0);
  static const hrtfForwardFallback = Vec3V1(0, 0, -1);
}

/// Hamilton storage order (w, x, y, z).
class QuatV1 {
  const QuatV1(this.w, this.x, this.y, this.z);

  final double w;
  final double x;
  final double y;
  final double z;

  static const identity = QuatV1(1, 0, 0, 0);

  double get norm => math.sqrt(w * w + x * x + y * y + z * z);

  bool get isDegenerate => norm < kQuatDegenerateNorm;

  QuatV1 normalized() {
    final n = norm;
    if (n < kQuatDegenerateNorm) return QuatV1.identity;
    return QuatV1(w / n, x / n, y / n, z / n);
  }

  QuatV1 get conjugate => QuatV1(w, -x, -y, -z);

  /// Inverse of the normalized quaternion (conjugate).
  QuatV1 inverse() => normalized().conjugate;
}

class SphericalV1 {
  const SphericalV1({
    required this.azimuthDeg,
    required this.elevationDeg,
    required this.distanceM,
  });

  final double azimuthDeg;
  final double elevationDeg;
  final double distanceM;
}

class ListenerPoseV1 {
  const ListenerPoseV1({
    required this.worldPosition,
    this.orientation = QuatV1.identity,
  });

  final Vec3V1 worldPosition;
  final QuatV1 orientation;
}

Vec3V1 _cross(Vec3V1 a, Vec3V1 b) => Vec3V1(
      a.y * b.z - a.z * b.y,
      a.z * b.x - a.x * b.z,
      a.x * b.y - a.y * b.x,
    );

/// Rotate [v] by unit quaternion [q] (listener-local → world when q is listener orientation).
Vec3V1 rotateVec3(QuatV1 q, Vec3V1 v) {
  final n = q.normalized();
  final qvec = Vec3V1(n.x, n.y, n.z);
  final t = _cross(qvec, v).scaled(2);
  return v + t.scaled(n.w) + _cross(qvec, t);
}

double wrapAzimuthDeg(double deg) {
  var az = deg;
  while (az > 180) {
    az -= 360;
  }
  while (az <= -180) {
    az += 360;
  }
  return az;
}

/// Signed shortest arc in [-180, +180] degrees.
double shortestAzimuthDeltaDeg(double fromDeg, double toDeg) {
  var d = toDeg - fromDeg;
  while (d > 180) {
    d -= 360;
  }
  while (d < -180) {
    d += 360;
  }
  return d;
}

double lerpAzimuthDeg(double fromDeg, double toDeg, double t) {
  return wrapAzimuthDeg(fromDeg + shortestAzimuthDeltaDeg(fromDeg, toDeg) * t);
}

Vec3V1 sphericalToLocal(SphericalV1 p) {
  final az = p.azimuthDeg * math.pi / 180.0;
  final el = p.elevationDeg * math.pi / 180.0;
  final d = p.distanceM;
  final ce = math.cos(el);
  return Vec3V1(
    d * math.sin(az) * ce,
    d * math.sin(el),
    -d * math.cos(az) * ce,
  );
}

SphericalV1 localToSpherical(Vec3V1 p) {
  final r = p.length;
  if (r < kZeroDistanceM) {
    return const SphericalV1(azimuthDeg: 0, elevationDeg: 0, distanceM: 0);
  }
  final h = math.sqrt(p.x * p.x + p.z * p.z);
  if (h < kHorizontalPoleM) {
    final el = p.y >= 0 ? 90.0 : -90.0;
    return SphericalV1(azimuthDeg: 0, elevationDeg: el, distanceM: r);
  }
  final el = math.asin((p.y / r).clamp(-1.0, 1.0)) * 180.0 / math.pi;
  final az = wrapAzimuthDeg(math.atan2(p.x, -p.z) * 180.0 / math.pi);
  return SphericalV1(azimuthDeg: az, elevationDeg: el, distanceM: r);
}

Vec3V1 hrtfUnitFromLocal(Vec3V1 local) {
  final r = local.length;
  if (r < kZeroDistanceM) return Vec3V1.hrtfForwardFallback;
  return local.scaled(1.0 / r);
}

Vec3V1 worldToListenerLocal(Vec3V1 world, ListenerPoseV1 listener) {
  final rel = world - listener.worldPosition;
  return rotateVec3(listener.orientation.inverse(), rel);
}

Vec3V1 listenerLocalToWorld(Vec3V1 local, ListenerPoseV1 listener) {
  return listener.worldPosition + rotateVec3(listener.orientation, local);
}

/// Geometric distance is independent of the HRTF unit vector.
double geometricDistanceM(Vec3V1 local) {
  final r = local.length;
  return r < kZeroDistanceM ? 0 : r;
}
